### DISCUSSION TITLE
eslint-plugin-mdx v2 causes OOM errors in VSCode

### DISCUSSION THREAD

**Discussion by mshick** [mdx-js/mdx#2071] [General] [answered]
I've been using `eslint-plugin-mdx` v1 with the VSCode eslint/mdx extensions for some time, so I know how it generally works. Ever since v2 was released I've been unable to use it on several projects due to OOM errors in VSCode like attached. 

I've created a minimal reproduction with a default `create-next-app` blog, just to have something more like a real world situation. I've set it up as the results of running `npx create-next-app`, following the set-up for the `dbaeumer.vscode-eslint`, `jounqin.vscode-mdx` and `eslint-plugin-mdx`. When I run eslint from the CLI it works as expected, with my prettier rules applied and errors correctly triggered. The eslint-driven formatting in VSCode even functions as expected until the OOMs cause the eslint server to shut down.

I can't help feel like I'm missing something, since no incompatibilities are noted on the relevant projects, nor has Google turned up anything. I don't have any other extensions enabled that affect MDX, as far as I can tell.

Here is my minimal repro: https://github.com/mshick/eslint-mdx-next

```sh
[Error - 2:22:48 PM] Connection to server got closed. Server will not be restarted.
[Info  - 2:22:48 PM] ESLint server stopped.
[Info  - 2:36:09 PM] ESLint server is starting
[Info  - 2:36:10 PM] ESLint server running in node v16.13.0
[Info  - 2:36:10 PM] ESLint server is running.
[Info  - 2:36:10 PM] ESLint library loaded from: /Users/mshick/Code/mshick/eslint-mdx-next/nextjs-blog/node_modules/eslint/lib/api.js

<--- Last few GCs --->

[91719:0x120798000]       44 ms: Mark-sweep (reduce) 1.4 (4.5) -> 1.4 (3.2) MB, 1.6 / 0.0 ms  (average mu = 0.076, current mu = 0.000) last resort GC in old space requested
[91719:0x120798000]       45 ms: Mark-sweep (reduce) 1.4 (3.2) -> 1.4 (3.0) MB, 1.0 / 0.0 ms  (average mu = 0.043, current mu = 0.000) last resort GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: 0x10d8bde6c node::FatalException(v8::Isolate*, v8::TryCatch const&) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 2: 0x10d8bdff4 node::OnFatalError(char const*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 3: 0x1095d6c44 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 4: 0x1095d6b74 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 5: 0x109788bf0 v8::internal::Heap::StartIncrementalMarking(int, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 6: 0x1097991b0 v8::internal::Heap::AllocatedExternalMemorySinceMarkCompact() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 7: 0x109762960 v8::internal::Factory::CodeBuilder::BuildInternal(bool) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 8: 0x109762b70 v8::internal::Factory::CodeBuilder::Build() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 9: 0x109d7a820 v8::internal::RegExpMacroAssemblerARM64::GetCode(v8::internal::Handle<v8::internal::String>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
10: 0x109b1d39c v8::internal::CharacterRange::IsCanonical(v8::internal::ZoneList<v8::internal::CharacterRange>*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
11: 0x109b452c8 v8::internal::RegExp::Exec(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, int, v8::internal::Handle<v8::internal::RegExpMatchInfo>, v8::internal::RegExp::ExecQuirks) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
12: 0x109b44a2c v8::internal::RegExp::Exec(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, int, v8::internal::Handle<v8::internal::RegExpMatchInfo>, v8::internal::RegExp::ExecQuirks) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
13: 0x109b42834 v8::internal::RegExp::Exec(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, int, v8::internal::Handle<v8::internal::RegExpMatchInfo>, v8::internal::RegExp::ExecQuirks) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
14: 0x109b9dff4 v8::internal::Runtime::DefineObjectOwnProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::StoreOrigin, v8::Maybe<v8::internal::ShouldThrow>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
15: 0x7407f0a06c 
16: 0x7407f6c8a0 
17: 0x7407e8d758 
18: 0x7407e8d758 
19: 0x7407e8d758 
20: 0x7407e8d758 
21: 0x7407e8d758 
22: 0x7407e8d758 
23: 0x7407e8d758 
24: 0x7407e8d758 
25: 0x7407e8d758 
26: 0x7407e8d758 
27: 0x7407e8d758 
28: 0x7407e8d758 
29: 0x7407e8d758 
30: 0x7407e8d758 
31: 0x7407e8d758 
32: 0x7407e8d758 
33: 0x7407e8b42c 
34: 0x7407e8b0c8 
35: 0x10970174c v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
36: 0x1095f4c68 v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
37: 0x10d82a084 node::CallbackScope::~CallbackScope() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
38: 0x10d83cbc0 node::EmitAsyncDestroy(node::Environment*, node::async_context) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
39: 0x10d903474 node::CommonEnvironmentSetup::context() const [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
40: 0x10824f3f0 uv_async_send [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
41: 0x108260214 uv_free_interface_addresses [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
42: 0x10824f80c uv_run [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
43: 0x10d8ff74c node::SpinEventLoop(node::Environment*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
44: 0x10d95ed48 node::PromiseRejectCallback(v8::PromiseRejectMessage) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
45: 0x10d961a20 node::PromiseRejectCallback(v8::PromiseRejectMessage) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
46: 0x1a5a5426c _pthread_start [/usr/lib/system/libsystem_pthread.dylib]
47: 0x1a5a4f08c thread_start [/usr/lib/system/libsystem_pthread.dylib]
[Info  - 2:36:18 PM] Connection to server got closed. Server will restart.
[Info  - 2:36:18 PM] ESLint server stopped.
[Info  - 2:36:18 PM] ESLint server is starting
[Info  - 2:36:18 PM] ESLint server running in node v16.13.0
[Info  - 2:36:18 PM] ESLint server is running.
[Info  - 2:36:18 PM] ESLint library loaded from: /Users/mshick/Code/mshick/eslint-mdx-next/nextjs-blog/node_modules/eslint/lib/api.js

<--- Last few GCs --->
```

**Comment by mshick** [comment]
Just to add a little more detail:

- I've disabled all VSCode extensions, restarted, then loaded the eslint and vscode-mdx extensions. The OOM errors begin right after enabling `vscode-mdx`. After another restart they persist.
- I've attempted switching to the pre-release version of the eslint extension, no difference.

Here is my VSCode info:

```
Version: 1.68.1
Commit: 30d9c6cd9483b2cc586687151bcbcd635f373630
Date: 2022-06-14T12:48:37.209Z (2 wks ago)
Electron: 17.4.7
Chromium: 98.0.4758.141
Node.js: 16.13.0
V8: 9.8.177.13-electron.0
OS: Darwin arm64 21.5.0
```

**Comment by JounQin** [comment, marked as answer]
So why not a `vscode-eslint` issue? `eslint` cli runs without any problem, right?

`eslint-mdx@v2` runs a worker inside through [`synckit`](https://github.com/un-ts/synckit).

Personally I don't have too much experience about VSCode extension.

---

See https://github.com/mdx-js/vscode-mdx#known-vscode-eslint-issues

**Reply by mshick** [reply]
@JounQin Are you suggesting I create an issue with the `vscode-eslint` project? Just to be clear, I only have issues when the `vscode-mdx` extension is enabled, and you appear to be the maintainer of that project.

**Reply by ChristianMurphy** [reply]
@mshick have you isolated it to just when `vscode-mdx` is enabled?
All your descriptions thus far (https://github.com/orgs/mdx-js/discussions/2071#discussion-4189564 and https://github.com/orgs/mdx-js/discussions/2071#discussioncomment-3072751) are predicated on both `vscode-eslint` AND `vscode-mdx` being enabled.

**Reply by mshick** [reply]
@ChristianMurphy Without `vscode-eslint` enabled the eslint server is never registered, so there is no crash. I have tested that configuration. I get the `mdx` syntax highlighting only in that scenario.

Going off the instructions for `vscode-mdx` it is generally treated as though it's an add-on for `vscode-eslint`, so I'm not sure how realistic it is to look at one without the other enabled. E.g., none of the `mdx` lint rules are applied or visible in VSCode in `.mdx` files without both extensions enabled. 

If the thinking is that `vscode-eslint` might have a more general issue with `synckit` plugins maybe you're aware of others I can test to file an issue against that repo?

**Reply by ChristianMurphy** [reply]
> Without vscode-eslint enabled the eslint server is never registered, so there is no crash. I have tested that configuration.

Right, which suggests the issue is with `vscode-eslint` and how it is interacting with `eslint-mdx`, rather than with the syntax highlighter `vscode-mdx`.

> Going off the instructions for `vscode-mdx` it is generally treated as though it's an add-on for `vscode-eslint`

It is not, is a syntax highlighter.
Which also happens to include some documentation, unrelated to its implementation, on how you can integrate vscode, mdx, and eslint together.

**Reply by ChristianMurphy** [reply]
> If the thinking is that `vscode-eslint` might have a more general issue with `synckit` plugins maybe you're aware of others I can test to file an issue against that repo?

It seems likely, and the replication you have (https://github.com/mshick/eslint-mdx-next) could be a good starting point.
Try removing `vscode-mdx` from the mix, and if the issue persists, than that would be something worth raising with https://github.com/microsoft/vscode-eslint to better understand what the incompatibility between it and `synckit` are.

**Reply by JounQin** [reply]
https://github.com/microsoft/vscode-eslint/issues/733#issuecomment-1005246167

Maybe this can be related. Please try to set a high memory for `vscode-eslint` runner.

**Reply by JounQin** [reply]
I raised https://github.com/microsoft/vscode-eslint/issues/1498 for tracking.

**Reply by JounQin** [reply]
@mshick See https://github.com/microsoft/vscode-eslint/issues/1498#issuecomment-1175813839

`"eslint.runtime": "node"` will do the trick.

**Reply by JounQin** [reply]
I'll add it into `vscode-mdx`'s document.

**Reply by JounQin** [reply]
> I'll add it into `vscode-mdx`'s document.

Done. https://github.com/mdx-js/vscode-mdx#known-vscode-eslint-issues

**Reply by mshick** [reply]
@JounQin Thanks for finding this. Working well for me.
